packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp

```diff
@@ -5442,7 +5449,14 @@ namespace Tpetra {
          std::runtime_error, "X and Y may not alias one another.");
     }
 
-    lclMatrix_->apply (X_lcl, Y_lcl, mode, alpha, beta);
+    LocalOrdinal nrows = getNodeNumRows();
+    LocalOrdinal maxRowImbalance = 0;
+    if(nrows != 0)
+      maxRowImbalance = getNodeMaxNumRowEntries() - (getNodeNumEntries() / nrows);
+    if(size_t(maxRowImbalance) >= Tpetra::Details::Behavior::rowImbalanceThreshold())
```

Comment:
A question which may not be very important. Here, to apply merge based algorithm we check load imbalance ratio to detect long rows. This means that we still use the so-called standard spmv for a balanced matrix of long rows. What I would like to point is that the merge based spmv does not only resolve the workload imbalance but it also creates more concurrency. For the balanced long  rows, it can still chop the long rows into multiple pieces (I am not 100% sure; but, the merge algorithm should be able to do so). Thus, more concurrent threads can work efficiently. On the other hand, the current kokkos spmv distributes threads over multiple rows (team-thread) and row reduction (thread-vector). This might be suboptimal choice for the balanced long row matrices. Even though we can increase the team size, it might not work well and under-utilize the device for a collection of extremely long rows.

Reply:
@kyungjoo-kim That's a good point. Now I'm thinking a simple "long rows" cutoff like before is what we want:
```
if getNodeMaxNumRowEntries() >= threshold:
  do_merge_path()
else
  do_normal_spmv()
```
Then, only if all rows are shorter than the threshold will normal SPMV be used. So the threshold would be chosen to be a little higher than the maximum degree seen in a regular sparse matrix.

Reply:
@kyungjoo-kim @kddevin  I did some experiments on V100 to test this case in KokkosKernels. What I found was that although KokkosKernels SPMV becomes ~2x slower for the few rows but many entries case, cuSPARSE's default algorithm actually stays slightly faster than merge path. It seems that merge path really is only beneficial for high imbalance.

Example that generates a fully dense matrix that's 100x100k.
```
[bmkelle@kokkos-dev-2 sparse]$ ./sparse_spmv_merge_wide --compare -numRows 100 -numCols 100000
Test parameters:
  - loop:           100
  - compare:        1
  - numRows:        100
  - numCols:        100000

Matrix statistics:
  - average nnz per row: 100000
cuSPARSE Merge alg    ---  min: 1.14594e-06 max: 2.27052e-06 avg: 1.16123e-06
cuSPARSE Default alg  ---  min: 1.08963e-06 max: 1.15345e-06 avg: 1.10374e-06
Kokkos Native alg     ---  min: 1.99581e-06 max: 2.0303e-06 avg: 2.00509e-06
```
More extreme: 10x1 million. 10 is way less than the number of SM's in the device, so no way KokkosKernels is getting good occupancy even at 1 team per row):
```
[bmkelle@kokkos-dev-2 sparse]$ ./sparse_spmv_merge_wide --compare -numRows 10 -numCols 1000000
Test parameters:
  - loop:           100
  - compare:        1
  - numRows:        10
  - numCols:        1000000

Matrix statistics:
  - average nnz per row: 1000000
cuSPARSE Merge alg    ---  min: 1.04111e-05 max: 2.22215e-05 avg: 1.06343e-05
cuSPARSE Default alg  ---  min: 1.04109e-05 max: 1.04592e-05 avg: 1.04225e-05
Kokkos Native alg     ---  min: 2.03733e-05 max: 2.04348e-05 avg: 2.03899e-05
```
It seems that cuSPARSE's default algorithm can make more than 1 block work on each row, unlike KokkosKernels. Although KokkosKernels isn't that bad even in this extreme case.

I think the conclusion to draw from this is that the current "imbalance" metric is a good choice, since whenever it's not getting triggered, the balance will be good enough for cuSPARSE's default to do well.

Reply:
Let me re-iterate abit. For the first example, it is dense gemv with long rows. A = (100x100k). Gemv flops is 2*100*100k/1e9 = 0.02 GFlop. This is done with 1.0e-6 second, which means 20 Tflops. Am I right ? This seems to be way faster than it could be in gemv. In terms of bandwidth, 100*100k*8 bytes = 0.08 GB are loaded. Within 1.0e-6 second, it gives 80 TB/s. Either I calculate something really stupid or the measurement does not seem to be right.

Reply:
@kyungjoo-kim Wow, you're totally right. For this experiment I made a copy of KokkosSparse_spmv_merge.cpp, in perf_test/sparse, and it seems to divide the times by 1000 at the end:
Measuring (this looks OK):
```
        for(int iterIdx = 0; iterIdx < loop; ++iterIdx) {
          Kokkos::Timer timer;
          // KokkosSparse::spmv(controls, "N", alpha, test_matrix, x, beta, y);
          KokkosSparse::Impl::spmv_beta<matrix_type, values_type, values_type, 1>(controls, "N", alpha, test_matrix, x, beta, y); 
          Kokkos::fence();
          double time = timer.seconds();
          avg_time += time;
          if(time>max_time) max_time = time;
          if(time<min_time) min_time = time;
        }   
```
Reporting:
```
        std::cout << "Kokkos Native alg     ---  min: " << min_time / 1000
                  << " max: " << max_time / 1000
                  << " avg: " << avg_time / (loop*1000) << std::endl;
```
I don't know why it does this. It's still valid to compare times against each other because they're all scaled by the same factor, but I will put in a PR to fix that.

Reply:
Now it makes sense. So, the number is 20 Gflop/s and 80 Gflop/s. This evaluates some aspects of the code. I agree that cusparse spmv select a different code path than a usual cases. Comparing the codes between nvidia spmv and nvidia merge spmv, it shows almost no difference. Actually, we do not know how nvidia spmv works. It may internally choose merge spmv or some specialized algorithm (changing the spmv into gemv) for this type of matrices. I generally assume that you test simulation a case that those long row column indices are contiguous. Am I right ? Are column indices sorted ? Is sorted column indices requirement for cusparse spmv ? I am just thinking if there is way that we can avoid the specialized code path in this evaluation or not.  

Since two algorithms perform almost same, i think you can choose any selection criteria whatever you like.

Reply:
@kyungjoo-kim The answers to your questions about the 10 x million and 10 x 100k matrices are yes (contiguous) yes (sorted), no (sorting required by cuSPARSE). If I modify the 10 x million matrix to randomly shuffle column indices within each row, I get this:
```
[bmkelle@kokkos-dev-2 sparse]$ ./sparse_spmv_merge_wide --compare -numRows 10 -numCols 1000000
Test parameters:
  - loop:           100
  - compare:        1
  - numRows:        10
  - numCols:        1000000

Matrix statistics:
  - average nnz per row: 1000000
cuSPARSE Merge alg    ---  min: 0.0129282 max: 0.0260412 avg: 0.013524
cuSPARSE Default alg  ---  min: 0.0119601 max: 0.0144351 avg: 0.0131062
Kokkos Native alg     ---  min: 0.0271301 max: 0.0445138 avg: 0.0300604
```
With the rows sorted and contiguous:
```
[bmkelle@kokkos-dev-2 sparse]$ ./sparse_spmv_merge_wide --compare -numRows 10 -numCols 1000000
Test parameters:
  - loop:           100
  - compare:        1
  - numRows:        10
  - numCols:        1000000

Matrix statistics:
  - average nnz per row: 1000000
cuSPARSE Merge alg    ---  min: 0.0104084 max: 0.0222391 avg: 0.010676
cuSPARSE Default alg  ---  min: 0.010407 max: 0.0114375 avg: 0.0104316
Kokkos Native alg     ---  min: 0.020373 max: 0.0204352 avg: 0.0203861
```
I'm kind of surprised the difference between shuffled and sorted is so small, only like 30-50%. But, the mergepath vs. default comparison doesn't change.

Reply:
Thank you for the comparison. Usually cusparse takes longer for the first iteration. The higher max time of the cusparse merge algorithm is probably due to the initial start-up overhead. Also, we see very close performance in the average time. With 80GB/s bandwidth utilization (the actual bandwidth utilization is probably higher than that), the GPU device is not fully utilized. So, judging algorithms with this 10x1M and 100x100k problems may not be right unless this problem size is meaningful for a certain sandia applications (even so, I would advise to use dense gemv for this case). Anyway I  think  that finding a practical use case for this type of problems would be difficult.

Reply:
@kyungjoo-kim I agree, I don't know of a use case for this shape of matrix. I was just trying to answer the question of whether merge path was better for the "balanced but long rows" case, where KokkosKernels (with at least 1 team per row) can't get good occupancy.

Reply:
@brian-kelley sorry for the misleading choice of units in my example, I probably thought it was useful at some point but definitely did not document it in anyways... I think when I worked on the SpMV the usual difficulty is to load the X vector efficiently and that leads to poor occupancy at least in the kk-native algorithm.
Regarding longer vs shorter rows, we usually try to handle that with the `vector_size` we use for the inner most reduction. This seems to be done similarly in cusparse at least according to the papers I read about it.